### PR TITLE
Cancelling suspend

### PR DESCRIPTION
**Pull request by Aidan63**
Suspension function which responds promptly to cancellation for you and gives you a callback to cancel / cleanup stuff. Delays implementation looks a lot nicer wtih this. Things still to do.

- I've not yet convinced myself the implementation is sound. I feel like all my continuation knowledge has been pushed out of my brain to make space for all the task knowledge. So I might just be unfamiliar with this stuff again.
- There are almost certainly more weird edge cases we can add tests for.
- Documentation
- I don't like the name but also can't come up with something better. `cancellingSuspend`, `suspendCancelling`, `promptSuspend`, are my best ideas and I think they're all naff.

### CODE REVIEW

**Comment by Simn** [comment]
I like this a lot! The way `delay` looks now makes me think this is the right approach.

> I've not yet convinced myself the implementation is sound. I feel like all my continuation knowledge has been pushed out of my brain to make space for all the task knowledge. So I might just be unfamiliar with this stuff again.

It's the opposite for me because the profiler has been pointing me towards continuation things, so I've been looking into that quite a bit. We can just take turns knowing about things.

For what it's worth, the approach here looks sound to me. The only thing that I think could happen is a race between a normal `resume` call and `doCancellation`, similar to what can happen in `RacingContinuation`. It might make sense to generalize the latter a little so that `CancellingContinuation` can extend it.

Another thing is that I've trained myself to accept one simple ground truth when it comes to continuation chains: whenever we call `resume` on something that doesn't belong to our current continuation, we should `schedule` that. This will ultimately allow a dispatcher to deal with it instead of making the call stack deeper.

And speaking of `RacingContinuation`, this makes me wonder if its own `resume` method should also check for cancellation. 

> I don't like the name but also can't come up with something better. cancellingSuspend, suspendCancelling, promptSuspend, are my best ideas and I think they're all naff.

I think Kotlin got this right with its "suspendCancellable" - we suspend something that is cancellable.

File: std/hxcoro/continuations/CancellingContinuation.hx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import haxe.coro.cancellation.CancellationToken;
+import haxe.Exception;
+import haxe.coro.context.Context;
+import haxe.coro.IContinuation;
+import haxe.coro.ICancellingContinuation;
+
+class CancellingContinuation<T> implements ICancellingContinuation<T> {
+	final cont : IContinuation<T>;
+
+	final handle : ICancellationHandle;
```
Comment (Simn) [inline std/hxcoro/continuations/CancellingContinuation.hx:25]: This is set in the constructor but never used.

Comment (Aidan63) [reply std/hxcoro/continuations/CancellingContinuation.hx:25]: Ah, right, that needs to be closed when resuming. I'm guessing if the suspend resumes normally but later the task is cancelled any registered callback will be ran and try and resume again!
Something to add a test for.

**Comment by Aidan63** [comment]
I've setup the continuation calls to go through the scheduler. I wasn't sure if closing the handle shoould happen inside that scheduled call as well, I've left it out for now.
I'll look at introducing the racing continuations race here as well, see if we can schedule some collisions with the virtual time scheduler as well.

**Comment by Simn** [comment]
I merged `kt_coro` without checking if it worked and then I couldn't immediately figure out how to adjust the `onCancellationRequested` call in `CoroSemaphore` to work with `cancellingSuspend`, so for now I just patched that in a weird way.

**Comment by Simn** [comment]
I don't want to be pedantic but "cancelling" sounds like something that is currently in the process of cancelling... I really think "cancellable" is more accurate, even though I hate actually saying the word.

**Comment by Aidan63** [comment]
no thats fair, I thought it was already called `cancellableSuspend` so just flipped the works. If I had bothered to actually read it I'd have seen that it was called `cancellingSuspend`.

**Comment by Aidan63** [comment]
I don't know what these failing tests are about, cpp and jvm all pass fine locally for me.

**Comment by Simn** [comment]
Hmm, no idea what's going on with the tests, it works for me locally as well...

**Comment by Simn** [comment]
This looks good to me now!

**Comment by Aidan63** [comment]
Don't think I've got anything else left to add either.
